lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -78,6 +79,14 @@ sub ajax {
             my @data;
             my %modules;
 
+            my %needles_record_real_path = map {
+                $_->path => {
+                    needle_file_real_path => realpath($_->path),
+                    last_seen             => $_->last_seen_time || 'never',
+                    last_match            => $_->last_matched_time || 'never',
+                }
+            } $needles->all;
```

Comment:
The duplication is worrysome. I don't understand why you also need to update here what is done later in the next while-loop. We can at least extract a function for the initialisation for last_seen+last_match. Maybe you have an even better idea.

Reply:
Yes. You could for instance create a simple mapping from path to DBIx object: `my %needles_record_real_path = map { $_->path => $_ }  $needles->all` and access the DBIx object directly in the loop below where you do the "never"-fallback only *one* time in the end like this:

```
$hash->{last_seen} ||= 'never';
$hash->{last_match} ||= 'never';
```

Reply:
I modified the code, please help to review if this way is better. Thanks